controllers/tykoasapidefinition_controller.go

```diff
@@ -188,6 +220,173 @@ func (r *TykOasApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.
+	log logr.Logger,
+	versionedTykOAS *v1alpha1.TykOasApiDefinition,
+) error {
+	if !versionedTykOAS.ObjectMeta.DeletionTimestamp.IsZero() {
+		if versionedTykOAS.Status.IsDefaultVersion {
+			errMsg := "Cannot delete default API version. Remove it from base API's default first"
+
+			versionedTykOAS.Status.LatestTransaction = v1alpha1.TransactionInfo{
+				Time:   metav1.Now(),
+				Status: v1alpha1.TransactionStatus(errMsg),
+			}
+
+			if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+				return err
+			}
+
+			return nil
+		}
+
+		var baseOasAPI v1alpha1.TykOasApiDefinition
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      versionedTykOAS.Status.BaseVersionName,
+			Namespace: versionedTykOAS.Status.BaseVersionNamespace,
+		}, &baseOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("failed to get TykOasApiDefinition CR", "err", err.Error())
+			}
+
+			return err
+		}
+
+		errMsg := "Cannot delete API. update the Base OasAPI to delete"
+
+		versionedTykOAS.Status.LatestTransaction = v1alpha1.TransactionInfo{
+			Time:   metav1.Now(),
+			Status: v1alpha1.TransactionStatus(errMsg),
+		}
+
+		if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+			return err
+		}
+	}
+
+	// this cleans up dangling versioned OAS APIs
+	if len(versionedTykOAS.Status.BaseVersionName) > 0 {
+		var baseOasAPI v1alpha1.TykOasApiDefinition
+
+		ns := versionedTykOAS.Status.BaseVersionNamespace
+		if ns == "" {
+			ns = "default"
+		}
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      versionedTykOAS.Status.BaseVersionName,
+			Namespace: ns,
+		}, &baseOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("base oas version not found, detaching version from the base OAS API.")
+
+				versionedTykOAS.RemoveOASVersionStatus()
+
+				if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+					return err
+				}
+
+				return nil
+			}
+
+			return err
+		}
+
+		var oasAPIPresent bool
+
+		for _, v := range baseOasAPI.Spec.Versioning.Versions {
+			if v.OasApiDefinitionRef == versionedTykOAS.Name {
+				oasAPIPresent = true
+			}
+		}
+
+		if !oasAPIPresent {
+			versionedTykOAS.RemoveOASVersionStatus()
+
+			if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+func (r *TykOasApiDefinitionReconciler) handleVersioningTykOASAPI(ctx context.Context,
+	tykBaseOASCrd *v1alpha1.TykOasApiDefinition,
+	log logr.Logger,
+	tykBaseOASDoc,
+	baseAPIID string,
+) error {
+	var versions []*TykOasApiDefinitionVersion
+
+	for _, v := range tykBaseOASCrd.Spec.Versioning.Versions {
+		var versionOasAPI v1alpha1.TykOasApiDefinition
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      v.OasApiDefinitionRef,
+			Namespace: v.Namespace,
+		}, &versionOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("failed to get versioned TykOasApiDefinition CR", "err", err.Error())
+			}
```

Comment:
why do we only print the log when the resource is not found on k8s? in my opinion, as we return from the loop whenever an error occurs, it'd be better to print the log whenever an error occurs. what do you think?

Reply:
@olamilekan000 i still see this is the case. can you please have a look once again? What do you think about this?

Reply:
```		
if err := r.Get(ctx, types.NamespacedName{
			Name:      v.TykOasApiDefinitionRef,
			Namespace: v.Namespace,
		}, &versionOasAPI); err != nil {
			if k8serrors.IsNotFound(err) {
				log.Info("failed to get versioned TykOasApiDefinition CR", "err", err.Error())
			}

			return err
		}
```

This is what the logic looks like, the log is just logging the specific error

Reply:
my question is why are we logging the specific error? can't we just print all errors?

Reply:
also, we are returning this error, so please use `log.Error`

Reply:
https://github.com/TykTechnologies/tyk-operator/pull/719/files#diff-6cd635972dc870184cc2eb7fd04f8ccf46d6821cbec2f194fb419469cc4ba437R456

it's used here already

Reply:
if it is printed, no need to re-print here as well. please ignore this log message